components/selection/selection-mixin.js

```diff
@@ -223,7 +226,24 @@ export const SelectionMixin = superclass => class extends RtlMixin(CollectionMix
 			const info = this.getSelectionInfo(true);
 			this._selectionObservers.forEach(observer => observer.selectionInfo = info);
 			this._updateObserversRequested = false;
-		}, 0);
+		});
+	}
+
+	_updateSortedSelectables() {
```

Comment:
This list could have a lot of items and would be called for each selection input that is subscribed or unsubscribed. The potential perf issue (mainly during initial load or load more) is partly mitigated by debouncing using the `setTimeout` here. If items are drag & dropped, the old item is removed from the DOM/unsubscribed, and a new item is created /subscribed so I _think_ DOM order mutations are covered by this. Having said that, this runs whether it's needed or not. Pondering options... 

We could wait to initialize this until the first time `_handleRadioKeyUp()` is invoked, the downside being that there could be a noticeable delay, which wouldn't be great. Or, we could do the sort upon subscription in a [requestIdleCallback](https://github.com/BrightspaceUI/core/blob/main/helpers/requestIdleCallback.js)(Safari still doesn't support it, hence our helper) if we wanted to be more zealous about deferring the work. I think it might be hard gauge the usefulness of that, because we're already deferring using `setTimeout`, and it's highly dependent on other work that's queued. I'm not certain the cost of this is impactful, but from previous experiments I did find that perf of large lists already degrades quite a bit (ex. quizzing 1000+ items with deep nesting), so being cautious here.

I was originally thinking we could avoid any upfront cost by creating `getNextElement(node, predicate, options)` and `getPreviousElement(node, predicate, options)` in our [dom.js](https://github.com/BrightspaceUI/core/blob/main/helpers/dom.js) helpers similar to [getNextFocusable](https://github.com/BrightspaceUI/core/blob/main/helpers/focus.js#L93) and [getPreviousFocusable](https://github.com/BrightspaceUI/core/blob/main/helpers/focus.js#L60) where the provided predicate would check for `d2l-selection-input`, that it belongs to the selection mixin's set, and is not `hidden` or `disabled`. Those focus helpers could perhaps be refactored to use them if appropriate, by moving the focusable check into a predicate. It would involve some "minimal" DOM walking, but only when a keyboard user is changing the selection. The DOM walking would be directly proportional to the amount of DOM the consumer is stuffing into the list-item, because it would typically be walking past those elements until it walks to the sibling list-item, and then into it to eventually land on the adjacent `d2l-selection-input`. The logic to do this pretty much already exists - I think we'd just need to make it more generic. 

I think it's jus a matter of what we prefer, and whether there are any gotcha's that we haven't thought of with these approaches.

Reply:
> I was originally thinking we could avoid any upfront cost by creating getNextElement(node, predicate, options) and getPreviousElement(node, predicate, options) in our [dom.js](https://github.com/BrightspaceUI/core/blob/main/helpers/dom.js) helpers...

One benefit of this would be then it definitely doesn't care about the subscriber stuff

Reply:
Another option might be to just call `getNextFocusable()`/`getPreviousFocusable()`. It could get a focusable that we aren't interested in, such as a secondary action or some other focusable element in the list-item content, so we could just call it in a loop until we locate a `d2l-selection-input` that meets our criteria. No refactoring of those helpers required. The generic helpers I mentioned above might be valuable in the future though.

Reply:
If I had to guess I'd think calling `compareDocumentPosition` on a series of direct siblings would be pretty fast to do up front, and less risky than doing almost any amount of tree walking in real-time on every key event. But maybe not. Shouldn't be too hard to test both out on a giant list.

It can definitely be limited to only the radio inputs, and maybe delayed to the first focus?

Reply:
With the helpers, the cost (both processing and memory, not that were talking _that_ big of a list) is the same regardless of the size of the list, whereas if we're sorting, the cost increases as the list gets bigger. I _think_ the sorting solution covers the mutation cases, but maybe we're missing something - waiting till they hit the arrow keys feels safer because it treats the DOM as the single source of truth and avoids mutation issues altogether, albeit at the risk of doing the tree-walking incorrectly. Having said that, those helpers are used a lot so I don't think there's a tonne of risk there (famous last words!). If it turns out that using the helpers is janky, I might change my preference though.